File: src/main/java/fun/vs/fw/demo/messages/TransactionKafkaListener.java
package fun.vs.fw.demo.messages;


import fun.vs.fw.demo.service.TransactionCategorizationService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.stereotype.Component;

/**
 * A listener class responsible for processing transaction messages received from a Kafka topic.
 * It uses the {@link TransactionCategorizationService} to categorize transactions and save categorized details.
 */
@Component
public class TransactionKafkaListener {

    private static final Logger log = LoggerFactory.getLogger(TransactionKafkaListener.class);
    private final TransactionCategorizationService categorizationService;

    public TransactionKafkaListener(TransactionCategorizationService categorizationService) {
        this.categorizationService = categorizationService;
    }

    /**
     * Listens to messages from the "transactions" Kafka topic and processes them for categorization.
     *
     * @param transaction The transaction message received from the "transactions" Kafka topic. This contains details
     *                    like transaction ID, account ID, client ID, merchant category code (MCC), and the transaction amount.
     */
    @KafkaListener(topics = "transactions", groupId = "transaction-group")
    public void listen(TransactionMessage transaction) {
        log.info("Received transaction: {}", transaction);
        categorizationService.categorizeTransaction(transaction);
    }
}

File: fvf4k/src/main/kotlin/fvf4k/demo/infra/kafka/TransactionKafkaListener.kt
package fvf4k.demo.infra.kafka

import arrow.core.raise.either
import fvf4k.demo.domain.api.CategorizeTransaction
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.kafka.annotation.KafkaListener

private val log = KotlinLogging.logger {}

class TransactionKafkaListener(val categorizeTransaction: CategorizeTransaction) {

    @KafkaListener(topics = ["transactions"], groupId = "fvf4k-transactions-group")
    fun listen(transaction: TransactionMessage) {
        log.info { "Received transaction $transaction" }
        either {
            val domainTransactionResult = either {
                transaction.toDomain()
            }
            
            domainTransactionResult.fold(
                ifLeft = { validationError ->
                    log.error { "Validation error: $validationError" }
                    return@listen
                },
                ifRight = { domainTransaction ->
                    categorizeTransaction(domainTransaction)
                }
            )
        }.onLeft { error ->
            log.error { "Error during categorization: $error" }
        }
    }
}